pkg/server/plugin/nodeattestor/tpmdevid/devid.go

```diff
@@ -0,0 +1,491 @@
+package tpmdevid
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/sha1" //nolint: gosec // SHA1 use is according to specification
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	common_devid "github.com/spiffe/spire/pkg/common/plugin/tpmdevid"
+	"github.com/spiffe/spire/pkg/common/util"
+	spc "github.com/spiffe/spire/proto/spire/common"
+	spi "github.com/spiffe/spire/proto/spire/common/plugin"
+	nodeattestorv0 "github.com/spiffe/spire/proto/spire/plugin/server/nodeattestor/v0"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func builtin(p *Plugin) catalog.BuiltIn {
+	return catalog.MakeBuiltIn(common_devid.PluginName,
+		nodeattestorv0.NodeAttestorPluginServer(p),
+	)
+}
+
+type config struct {
+	trustDomain string
+
+	devIDRoots          *x509.CertPool
+	ekRoots             *x509.CertPool
+	checkDevIDResidency bool
+}
+
+type Config struct {
+	DevIDBundlePath       string `hcl:"devid_bundle_path"`
+	EndorsementBundlePath string `hcl:"endorsement_bundle_path"`
+	CheckDevIDResidency   bool   `hcl:"check_devid_residency"`
+}
+
+type Plugin struct {
+	nodeattestorv0.UnsafeNodeAttestorServer
+
+	m sync.Mutex
+	c *config
+}
+
+func New() *Plugin {
+	return &Plugin{}
+}
+
+func (p *Plugin) Attest(stream nodeattestorv0.NodeAttestor_AttestServer) error {
+	// Receive attestation request
+	req, err := stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	conf := p.getConfiguration()
+	if conf == nil {
+		return status.Error(codes.FailedPrecondition, "not configured")
+	}
+
+	if dataType := req.AttestationData.Type; dataType != common_devid.PluginName {
+		return status.Errorf(codes.InvalidArgument, "unexpected attestation data type %q", dataType)
+	}
+
+	// Unmarshall received attestation data
+	attData := new(common_devid.AttestationRequest)
+	err = json.Unmarshal(req.AttestationData.Data, attData)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to unmarshall attestation data: %v", err)
+	}
+
+	// Decode attestation data
+	if len(attData.DevIDCert) == 0 {
+		return status.Error(codes.InvalidArgument, "no DevID certificate to attest")
+	}
+
+	devIDCert, err := x509.ParseCertificate(attData.DevIDCert)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "unable to parse DevID certificate: %v", err)
+	}
+
+	// Verify DevID certificate chain of trust
+	err = verifyDevIDSignature(devIDCert, conf.devIDRoots)
+	if err != nil {
+		return status.Errorf(codes.Unauthenticated, "unable to verify DevID signature: %v", err)
```

Comment:
I'm not sure if this code actually bubbles back to the caller or not ... it might be caught and processed in SPIRE Server core. The unauthenticated code may also be a little misleading, because the API the caller is interacting with isn't technically an authenticated API ... so I think that `codes.InvalidArgument` or something similar may be more appropriate here